app/services/foreman/provision/ssh.rb

```diff
@@ -50,64 +78,62 @@ def command_prefix
+  # tell if other locations are writeable or executable by the user.
+  def main_command
     main_execution = "(chmod 0701 ./#{remote_script} && #{command_prefix} ./#{remote_script} ; echo $? >#{remote_script}.status) 2>&1"
     "#{command_prefix} sh -c '#{main_execution} | tee #{remote_script}.log; exit $(cat #{remote_script}.status)'"
   end
 
-  def defaults
-    {
-      :keys_only    => true,
-      :config       => false,
-      :auth_methods => %w(publickey),
-      :compression  => true,
-      :logger       => logger,
-    }
-  end
-
   def logger
     Rails.logger
   end
 
-  def initiate_connection!
-    Timeout.timeout(Setting[:ssh_timeout].to_i) do
-      Timeout.timeout(8) do
-        ssh.run('pwd')
-      end
-    rescue Errno::ECONNREFUSED
-      logger.debug "Connection refused for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Errno::EHOSTUNREACH
-      logger.debug "Host unreachable for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::Disconnect
-      logger.debug "Host dropping connections for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::ConnectionTimeout
-      logger.debug "Host timed out for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::AuthenticationFailed
-      logger.debug "Auth failed for #{username} at #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Timeout::Error
-      retry
-    rescue => e
-      Foreman::Logging.exception("SSH error", e)
-    end
+  def scp(local)
+    logger.debug "about to upload #{template} to remote system at #{remote_script}"
+    logger.debug scp_cmd(local)
+    _stdout, stderr, status = Open3.capture3(scp_cmd(local))
+
+    return if status.success?
+
+    logger.error stderr
+    raise Foreman::Exception.new("Failed to upload '#{local}' to remote system at #{address}")
+  end
+
+  def passwd_cmd
+    return unless ssh_auth_method == :password
+
+    "sshpass -p '#{@options[:password]}'"
   end
 
-  def ssh
-    Fog::SSH.new(address, username, options.merge(:timeout => 4))
+  def ssh_key
+    return unless ssh_auth_method == :key
+    "-i #{tmp_key.path}"
+  end
+
+  def ssh_cmd(host_command)
+    parts = [passwd_cmd, "ssh", ssh_key, ssh_options, remote_host, "\"#{host_command}\""]
+    parts.compact.join(" ")
+  end
+
+  def scp_cmd(local)
+    parts = [passwd_cmd, "scp", ssh_key, ssh_options, local, "#{remote_host}:~/#{remote_script}"]
+    parts.compact.join(" ")
+  end
+
+  def ssh_options
+    opts = ["-o ConnectTimeout=6", "StrictHostKeyChecking=accept-new", "Compression=yes"]
+
+    opts << "PubkeyAuthentication=yes" if ssh_auth_method == :key
+    opts << "PasswordAuthentication=yes" if ssh_auth_method == :password
+
+    opts.join(" -o ")
   end
 
-  def scp
-    Fog::SCP.new(address, username, options)
+  def tmp_key
+    file = Tempfile.new("ssh-key-#{uuid}")
+    file.write(@options[:key_data].join("\n"))
```

Comment:
Why is `key_data` an array? Can ssh private keys even be concatenated into a single file?

Reply:
It's because of https://github.com/stejskalleos/foreman/blob/ls/ssh_sys/app/models/concerns/orchestration/ssh_provision.rb#L46 where it's an array.

> Can ssh private keys even be concatenated into a single file?

I was wondering by myself. I think it's safe to change it to a string instead of an array